sound/soc/sof/intel/hda-dai-ops.c

```diff
@@ -0,0 +1,413 @@
+	default:
+		return 0;
+	}
+
+	return 0;
+}
+
+static struct snd_sof_dai_dma_ops hda_dma_ops = {
+	.stream_get_dma = hda_stream_get_dma,
+	.stream_put_dma = hda_stream_put_dma,
+	.setup = hda_dma_setup,
+	.free = hda_dma_free,
+	.trigger = hda_trigger,
+};
+
+static int hda_codec_dai_params(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+				void *dma_data, struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params, unsigned int *val)
+{
+	struct hdac_ext_stream *hext_stream = dma_data;
+	struct hdac_ext_link *hlink;
+	struct hdac_bus *bus;
+	unsigned char stream_tag;
+	unsigned int link_bps;
+
+	bus = sof_to_bus(sdev);
+	hlink = snd_hdac_ext_bus_get_hlink_by_name(bus, codec_dai->component->name);
+	if (!hlink)
+		return -EINVAL;
+
+	stream_tag = hdac_stream(hext_stream)->stream_tag;
+	if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_PLAYBACK)
+		snd_hdac_ext_bus_link_set_stream_id(hlink, stream_tag);
+
+	/* set the hdac_stream in the codec dai */
+	snd_soc_dai_set_stream(codec_dai, hdac_stream(hext_stream), substream->stream);
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		link_bps = codec_dai->driver->playback.sig_bits;
+	else
+		link_bps = codec_dai->driver->capture.sig_bits;
+
+	*val = snd_hdac_calc_stream_format(params_rate(params), params_channels(params),
+					   params_format(params), link_bps, 0);
+
+	dev_dbg(bus->dev, "format_val=%d, rate=%d, ch=%d, format=%d\n",
+		*val, params_rate(params), params_channels(params), params_format(params));
+
+	return 0;
+}
+
+static int hda_codec_dai_free(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+			      void *dma_data, struct snd_pcm_substream *substream)
+{
+	struct hdac_ext_stream *hext_stream = dma_data;
+	struct hdac_ext_link *hlink;
+	struct hdac_bus *bus;
+
+	bus = sof_to_bus(sdev);
+	hlink = snd_hdac_ext_bus_get_hlink_by_name(bus, codec_dai->component->name);
+	if (!hlink)
+		return -EINVAL;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		unsigned char stream_tag;
+
+		stream_tag = hdac_stream(hext_stream)->stream_tag;
+		snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag);
+	}
+
+	return 0;
+}
+
+static struct snd_sof_dai_codec_ops hda_codec_ops = {
+	.params = hda_codec_dai_params,
+	.free = hda_codec_dai_free,
+};
+
+static const struct snd_sof_dai_ops hda_dai_ops = {
+	.dma_ops = &hda_dma_ops,
+	.codec_ops = &hda_codec_ops,
+};
+
+#endif
+
+void hda_set_dai_widget_ops(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget)
+{
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA_AUDIO_CODEC)
+	struct snd_sof_dai *sdai = swidget->private;
+
+	switch (sdev->pdata->ipc_type) {
+	case SOF_IPC:
+	{
+		struct sof_dai_private_data *private = sdai->private;
+		struct snd_sof_dai_dma_ops *dma_ops;
+
+		if (private->dai_config->type == SOF_DAI_INTEL_HDA) {
+			sdai->ops = &hda_dai_ops;
+			dma_ops = sdai->ops->dma_ops;
+
+			/* set the post DMA trigger op */
+			dma_ops->post_trigger = hda_ipc3_post_trigger;
```

Comment:
Why post? If you have one of the three trigger ops used it does not really matter which one you pick they are going to be in sequence and if there is only one used then it is the `trigger` itself as post would imply that there is something which precedes this call.

Reply:
so you'd get here for the case where you have paused a stream but the system suspends. The PAUSE trigger is noth IPC3 and IPC4 would have stopped the DMA. What we wouldnt have done is the post_trigger() is to free the DAI CONFIG for IPC3 and reset the pipeline for IPC4. And post_trigger() handles that

Reply:
comments needed here!